src/main/java/com/iota/iri/TransactionValidator.java

```diff
@@ -108,36 +124,37 @@ public static TransactionViewModel validate(final byte[] bytes, int minWeightMag
 
     private final AtomicInteger nextSubSolidGroup = new AtomicInteger(1);
 
-    public boolean checkSolidity(Hash hash, boolean milestone) throws Exception {
-        if(TransactionViewModel.fromHash(tangle, hash).isSolid()) {
+    public boolean checkSolidity(Hash hash, boolean priority) throws Exception {
+        if(TransactionViewModel.fromHash(tangle, hash).subtangleStatus() == SOLID) {
             return true;
         }
         Set<Hash> analyzedHashes = new HashSet<>(Collections.singleton(Hash.NULL_HASH));
-        boolean solid = true;
         final Queue<Hash> nonAnalyzedTransactions = new LinkedList<>(Collections.singleton(hash));
         Hash hashPointer;
         while ((hashPointer = nonAnalyzedTransactions.poll()) != null) {
             if (analyzedHashes.add(hashPointer)) {
                 final TransactionViewModel transaction = TransactionViewModel.fromHash(tangle, hashPointer);
-                if(!transaction.isSolid()) {
-                    if (transaction.getType() == TransactionViewModel.PREFILLED_SLOT && !hashPointer.equals(Hash.NULL_HASH)) {
-                        transactionRequester.requestTransaction(hashPointer, milestone);
-                        solid = false;
-                        break;
-                    } else {
-                        if (solid) {
+                switch (transaction.subtangleStatus()) {
+                    case INVALID:
+                        if(!priority) {
+                            propagateInvalidSubtangle(transaction.getHash(), priority);
+                            return false;
+                        }
```

Comment:
the lack of a `break` is used implicitly to have `priority` treat INVALID as UNKNOWN?
if so, "explicit is better than implicit" - may seem as a bug, not intentional opt.